contrib_bots/lib/giveaway.py

```diff
@@ -0,0 +1,215 @@
+# See readme.md for instructions on running this code.
+from __future__ import print_function
+import time
+import threading
+import random
+from past.utils import old_div
+
+class GiveawayHandler(object):
+    '''
+    This plugin allows for "giveaways" to be run within a
+    zulip stream. The bot will look for any message starting
+    with "@giveaway" followed by a command (either init, enter,
+    exit or cancel) followed by a time in seconds for the giveaway
+    to run for. Full syntax is shown below:
+    @giveaway command giveaway_name giveaway_time
+
+    Possible commands:
+    init - initialize a giveaway (only one can be running at a time)
+    enter - enter a giveaway (a giveaway must be running to use this command)
+    exit - leave a giveaway (must be partaking in an active giveaway for this
+    command to work
+    cancel - cancel a giveaway (can only be run by the person who first
+    initialised the giveaway
+
+    This plugin was created by Daniel O'Brien
+    as part of Google Code-In 2016-2017
+    '''
+    def __init__(self):
+        self.giveaway_array = []
+        self.giveaway_on = False
+        self.giveaway_name = ""
+        self.giveaway_time = 0
+        self.giveaway_creator = ""
+        self.giveaway_start_time = 0
+        self.giveaway_group = ""
+        self.giveaway_subject = ""
+        self.giveaway_client = ""
+        self.giveaway_timeMins = 0
```

Comment:
@umairwaheed sorry, didn't change that comment fully to use _ will change it now